### PR TITLE
Linked list

### PR DESCRIPTION
**Pull request by CCallahanIV**
First submission

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+from setuptools import setup
+
+setup(
+    name="linked_list",
```
Comment (nhuntwalker) [inline setup.py]: This repository will be for more than just the Linked List. It'll be for all of your data structures and algorithms

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    description="Week 2 Day 1 Assignment - Linked List",
```
Comment (nhuntwalker) [inline setup.py]: Again, more than just the Linked list

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    version=0.0,
```
Comment (nhuntwalker) [inline setup.py]: You can give it a version number that's not 0.0

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    author="Ted Callahan and Mark Kessler-Wenicker",
+    author_email="",
+    license="MIT",
+    package_dir={'': 'src'},
+    py_modules=["linkedlist"],
```
Comment (nhuntwalker) [inline setup.py:11]: be sure to grow this list as your number of data structures grows

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    install_requires=[],
+    extras_require={"test": ["pytest", "pytest-watch", "pytest-cov", "tox"]},
+    entry_points={
+        'console_scripts': [
+            "linked_list = linked_list:main"
```
Comment (nhuntwalker) [inline setup.py]: no `main` function exists in `linked_list`

File: src/linked_list.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
+"""Creating a Linked List Data Structure.  CF 401 Python Week 2 assignment."""
+
+
+class Node(object):
+    """Create Node objects for use in a Linked List data structure."""
+    def __init__(self, value=None, nxt=None):
```
Comment (nhuntwalker) [inline src/linked_list.py:12]: You can put a doc string here too.

File: src/linked_list.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        self.nxt = nxt
+
+
+class Linked_List(object):
+    """Create a Linked List Data Structure."""
+    def __init__(self, iterable=None):
+        """Intialize a Linked List Object."""
+        self.head = None
+        self._size = 0
+        if iterable and hasattr(iterable, "__iter__"):
```
Comment (nhuntwalker) [inline src/linked_list.py]: Note, and something I neglected during lecture, strings in Python 2 don't have the `__iter__` attribute. You may want to extend this with:
```python
if iterable and hasattr(iterable, "__iter__") or isinstance(iterable, str):
    # do stuff
```

File: src/linked_list.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        return self._size
+
+    def pop(self):
+        """Pop the first value off the head of LL and return it."""
+        if self.head is None:
+            raise IndexError("Cannot pop from an empty Linked List.")
+        val = self.head.value
+        self.head = self.head.nxt
+        self._size -= 1
+
```
Comment (nhuntwalker) [inline src/linked_list.py]: Unnecessary white space

File: src/linked_list.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        return val
+
+    def search(self, query):
+        """Return the node that contains the value."""
+        match = [node for node in self._iterate_from(self.head) if node.value == query]
+        return match[0] if match else "That value is not in this linked list."
```
Comment (nhuntwalker) [inline src/linked_list.py]: Actually this is supposed to return `None` if no nodes in the list contain the value.

File: src/linked_list.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+    def display(self):
+        """Return a unicode string representing the Linked List as a tuple."""
+        return str(tuple(self._node_values()))
+
+    def __len__(self):
+        """Enable use of len() function."""
+        return self.size()
+
+    def _iterate_from(self, list_item):
```
Comment (nhuntwalker) [inline src/linked_list.py:105]: Um wow. You used a generator. Bravo!

File: src/linked_list.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        """Remove the given node from the LL."""
+        cur_node = self.head
+        if self.head is r_node:
+            self.head = self.head.nxt
+            self._size -= 1
+            return "Succesfully removed Node with value: \
+            {0}. New head set to {1}".format(r_node.value, self.head.value)
+        while cur_node:
+            if r_node == cur_node.nxt:
+                cur_node.nxt = cur_node.nxt.nxt
```
Comment (nhuntwalker) [inline src/linked_list.py]: What if the node you're looking for is the last node? `cur_node.nxt` would be `None`. Then `cur_node.nxt.nxt` would raise an `AttributeError`

File: src/linked_list.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+    def _iterate_from(self, list_item):
+        while list_item is not None:
+            yield list_item
+            list_item = list_item.nxt
+
+    def _node_values(self):
+        """Helper function to return an iterable of node values"""
+        node_values = [node.value for node in self._iterate_from(self.head)]
+        return node_values
```
Comment (nhuntwalker) [inline src/linked_list.py:113]: You could've folded this into the `display` method

File: src/test_linked_list.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    #result.startswith("Success")
+
+
+def test_linked_list_remove_node_not_exists():
+    """Test if the remove method of the Linked_List class is correctly behaving
+    if the node does not exist"""
+    from linked_list import Linked_List
+    ll = Linked_List(PARAMS_SAMPLE_LIST)
+    with pytest.raises(ValueError):
+        r_node = Node("bear")
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: `NameError` because you don't import the `Node` class in this test

File: src/test_stack.py [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+def test_node_init():
+    """Test the initialization of a node."""
+    from linked_list import Node
+    assert type(Node) == type
+
+
+def test_node_instantiation():
+    """Test if the Node objects gets initiated correctly"""
+    from linked_list import Node, init_node
```
Comment (nhuntwalker) [inline src/test_stack.py]: `ImportError` because there's no `init_node` function in the `linked_list` module

File: src/test_linked_list.py [hunk 15]
```diff
@@ -0,0 +1,127 @@
+"""Testing the linked_list class - CF 401 Python Week 2 Assignment."""
+import pytest
+
+PARAMS_SAMPLE_LIST = ["something", 1, "pear", 3, "apple"]
+
+@pytest.fixture
+def init_node():
+    from linked_list import Node
+    node = Node(0)
+    return str(type(node))
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: You never use this fixture.

File: src/test_linked_list.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+def create_empty_node():
+    from linked_list import Node
+    return Node()
+
+@pytest.fixture
+def create_ll_with_list():
+    """Given a list, create a new Linked List with all the values."""
+    from linked_list import Linked_List
+    new_ll = Linked_List(PARAMS_SAMPLE_LIST)
+    return new_ll
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: This fixture is never used

File: src/test_linked_list.py [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+@pytest.fixture
+def create_empty_ll():
+    from linked_list import Linked_List
+    new_ll = Linked_List()
+    return new_ll
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: This fixture is never used

File: src/test_linked_list.py [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+def test_node_instantiation(create_empty_node):
+    """Test if the Node objects gets initiated correctly"""
+    assert create_empty_node.value is None
+
+
+def test_node_value():
+    """Test the contents of nodes created with given input."""
+    from linked_list import Node
+    node2 = Node("Something")
+    assert node2.value == "Something" and node2.nxt == None
```
Comment (nhuntwalker) [inline src/test_linked_list.py:51]: This would be better as `node2.value == "Something" and node2.nxt is None`

File: src/test_linked_list.py [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+def create_empty_ll():
+    from linked_list import Linked_List
+    new_ll = Linked_List()
+    return new_ll
+
+
+def test_node_init():
+    """Test the initialization of a node."""
+    from linked_list import Node
+    assert type(Node) == type
```
Comment (nhuntwalker) [inline src/test_linked_list.py:38]: All class objects are `type` objects. You don't need to test for that, that's handled by Python.

File: src/test_linked_list.py [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+    result = new_linked_list.search('pear')
+    assert result.value == 'pear'
+
+
+def test_linked_list_search_failure():
+    """Test the search method of the Linked List class with a search query that DOES NOT exist."""
+    from linked_list import Linked_List
+    new_linked_list = Linked_List(PARAMS_SAMPLE_LIST)
+    result = new_linked_list.search('owt')
+    assert result.startswith('That')
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: It should return `None` if it doesn't exist

File: src/test_linked_list.py [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+
+
+def test_linked_list_pop():
+    """Test the pop method of the Linked List class."""
+    from linked_list import Linked_List
+    new_ll = Linked_List(PARAMS_SAMPLE_LIST)
+    old_head_val = new_ll.head.value
+    new_ll.pop()
+    new_head_val = new_ll.head.value
+    assert old_head_val == "apple" and new_head_val == 3
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: All this tests is that the pop method reassigns the head. It doesn't test that the `pop` method returns a value at all, let alone that it returns the value that was at `head`

File: src/test_linked_list.py [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+
+def test_linked_list_size():
+    """Test if the size method of the Linked_List class returns the size
+    correctly by independently iterating over the the linked list"""
+    from linked_list import Linked_List
+    test_l = len(PARAMS_SAMPLE_LIST)
+    new_linked_list = Linked_List(PARAMS_SAMPLE_LIST)
+    assert new_linked_list._size == test_l
+    assert len(new_linked_list) == test_l
+    assert new_linked_list.size() == test_l
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: If `__len__` calls `size()` and `size() returns ._size` you're really only testing one thing.

File: src/test_linked_list.py [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+
+def test_linked_list_remove_node_exists():
+    """Test if the remove method of the Linked_List class is correctly
+    removing a node"""
+    from linked_list import Linked_List
+    ll = Linked_List(PARAMS_SAMPLE_LIST)
+    r_node = ll.search("pear")
+    ll.remove(r_node)
+    assert ll.size() == 4
+    #result.startswith("Success")
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: Remove commented code if you're not gonna use it.

File: src/test_linked_list.py [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+
+
+def test_linked_list_remove_node_exists():
+    """Test if the remove method of the Linked_List class is correctly
+    removing a node"""
+    from linked_list import Linked_List
+    ll = Linked_List(PARAMS_SAMPLE_LIST)
+    r_node = ll.search("pear")
+    ll.remove(r_node)
+    assert ll.size() == 4
```
Comment (nhuntwalker) [inline src/test_linked_list.py]: why hard code the length when you can just say `ll.size() == len(PARAMS_SAMPLE_LIST)`?

File: tox.ini [hunk 25]
```diff
@@ -0,0 +1,9 @@
+[tox]
+envlist = py27, py35
+
+[testenv]
+commands = py.test --cov-report term-missing
```
Comment (nhuntwalker) [inline tox.ini]: For this to actually return coverage reports in Py 2.7 and Py 3.5, this line should be
`commands = py.test src --cov=src --cov-report term-missing`

When I add coverage proper, I'm getting 88% coverage on your `linked_list` module.

File: src/linked_list.py [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+    """Create a Linked List Data Structure."""
+    def __init__(self, iterable=None):
+        """Intialize a Linked List Object."""
+        self.head = None
+        self._size = 0
+        if iterable and hasattr(iterable, "__iter__"):
+            for value in iterable:
+                self.push(value)
+        elif iterable:
+            raise TypeError
```
Comment (nhuntwalker) [inline src/linked_list.py]: This isn't tested

File: src/linked_list.py [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+        self._size += 1
+
+    def size(self):
+        """Return the length of the Linked_List."""
+        return self._size
+
+    def pop(self):
+        """Pop the first value off the head of LL and return it."""
+        if self.head is None:
+            raise IndexError("Cannot pop from an empty Linked List.")
```
Comment (nhuntwalker) [inline src/linked_list.py:64]: Untested

File: src/linked_list.py [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+        return match[0] if match else "That value is not in this linked list."
+
+    def remove(self, r_node):
+        """Remove the given node from the LL."""
+        cur_node = self.head
+        if self.head is r_node:
+            self.head = self.head.nxt
+            self._size -= 1
+            return "Succesfully removed Node with value: \
+            {0}. New head set to {1}".format(r_node.value, self.head.value)
```
Comment (nhuntwalker) [inline src/linked_list.py]: Untested

File: src/linked_list.py [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+        while cur_node:
+            if r_node == cur_node.nxt:
+                cur_node.nxt = cur_node.nxt.nxt
+                self._size -= 1
+                return "Succesfully removed Node with value: {}".format(r_node.value)
+            else:
+                cur_node = cur_node.nxt
+        else:
+            raise ValueError("ERROR: That node is not in this linked list.")
```
Comment (nhuntwalker) [inline src/linked_list.py:79]: Untested.

File: src/test_linked_list.py (hunk 19 above)
Comment (wenima) [reply src/test_linked_list.py:38]: tracked by https://github.com/wenima/data-structures/issues/4

File: src/test_linked_list.py (hunk 18 above)
Comment (wenima) [reply src/test_linked_list.py:51]: tracked by https://github.com/wenima/data-structures/issues/5

File: src/linked_list.py (hunk 27 above)
Comment (wenima) [reply src/linked_list.py:64]: tracked by https://github.com/wenima/data-structures/issues/6

File: setup.py (hunk 4 above)
Comment (wenima) [reply setup.py:11]: done

File: src/linked_list.py (hunk 6 above)
Comment (wenima) [reply src/linked_list.py:12]: tracked in https://github.com/wenima/data-structures/issues/7

File: src/linked_list.py [hunk 30]
```diff
@@ -0,0 +1,113 @@
...
+
+    search():       Given a Node value to search by, return the Node from the
+                    Linked List containing that value.  Else, return None.
+
+    remove():       Given a Node, remove that Node from the Linked List.  Else,
+                    raise a ValueError.
+
+    display():      Return a string of all Nodes in the Linked List,
+                    formatted as a Tuple.
+    """
```
Comment (nhuntwalker) [inline src/linked_list.py:40]: Loving the verbose documentation

**Pull request merged**